man/mhbuild.man
.TH MHBUILD %manext1% 2016-10-15 "%nmhversion%"
.
.\" %nmhwarning%
.
.SH NAME
mhbuild \- translate MIME composition drafts for nmh messages
.SH SYNOPSIS
.na
.HP 5
.B mhbuild
.RB [ \-help ]
.RB [ \-version ]
.I file
.RB [ \-auto " | " \-noauto ]
.RB [ \-list " | " \-nolist ]
.RB [ \-realsize " | " \-norealsize ]
.RB [ \-headers " | " \-noheaders ]
.RB [ \-directives " | " \-nodirectives ]
.RB [ \-rfc934mode " | " \-norfc934mode ]
.RB [ \-contentid " | " \-nocontentid ]
.RB [ \-verbose " | " \-noverbose ]
.RB [ \-disposition " | " \-nodisposition ]
.RB [ \-headerencoding
.I encoding-algorithm
.RB " | " \-autoheaderencoding ]
.RB [ \-maxunencoded
.IR line-length ]
.RB [ \-dist ]
.ad
.SH DESCRIPTION
The
.B mhbuild
command will translate a MIME composition draft into a valid MIME message.
.PP
.B mhbuild
creates multi-media messages as specified in RFCs 2045 through 2049.
This includes the encoding of message headers as specified by RFC 2047,
and, additionally, the encoding of MIME parameters as specified in RFC 2231.
.PP
If you specify the name of the composition file as \*(lq-\*(rq, then
.B mhbuild
will accept the composition draft on the standard input.
If the translation of this input is successful,
.B mhbuild
will output the new MIME message to the standard output.
This argument must be the last argument on the command line.
.PP
Otherwise, if the file argument to
.B mhbuild
is the name of a valid composition file, and the translation is successful,
.B mhbuild
will replace the original file with the new MIME message.
It will rename the original file to start with the \*(lq,\*(rq character
and end with the string \*(lq.orig\*(rq, e.g., if you are editing the file
\*(lqdraft\*(rq, it will be renamed to \*(lq,draft.orig\*(rq.
This allows you to easily recover the
.B mhbuild
input file.
.SS "Listing the Contents"
The
.B \-list
switch tells
.B mhbuild
to list the table of contents associated with the MIME message that is created.
.PP
The
.B \-headers
switch indicates that a one-line banner should be displayed above the listing.
The
.B \-realsize
switch tells
.B mhbuild
to evaluate the \*(lqnative\*(rq (decoded) format of each content prior to listing.
This provides an accurate count at the expense of a small delay.
If the
.B \-verbose
switch
is present, then the listing will show any \*(lqextra\*(rq information
that is present in the message, such as comments in the
\*(lqContent-Type\*(rq header.
.PP
If the
.B \-disposition
switch is present, then the listing will show any relevant information from
the \*(lqContent-Disposition\*(rq header.
.SS "Simplified Attachment Interface"
For users who wish to simply attach files to text content,
.B mhbuild
will scan the composition file for \*(lqAttach\*(rq headers.
An \*(lqAttach\*(rq header contains a filename that will be appended to the
message using normal MIME encapsulation rules.  One filename is allowed
per \*(lqAttach\*(rq header, but multiple \*(lqAttach\*(rq headers are
allowed per composition file.
.PP
These files will be appended after any other MIME content, including any
content specified by
.B mhbuild
directives (see below).  See
.IR send (1)
for more details.
.PP
By default, the Content-Disposition will be \*(lqattachment\*(rq.
.B mhbuild
looks for user profile and
.I mhn.defaults
entries of the form
.PP
.RS 5
.BI mhbuild-disposition- type / subtype
.RE
or
.RS 5
.BI mhbuild-disposition- type
.RE
.PP
to supply the disposition value.  The only supported values are
\*(lqattachment\*(rq
and
\*(lqinline\*(rq.
.SS "Convert Interface"
The convert interface is a powerful mechanism that supports
replying to MIME messages.
These placeholders are used in the following description:
.RS 5
.TP 15
.PD 0
TYPE
content type/subtype
.TP
CONVERTER
external program, and any fixed arguments, to convert content, such as
from a request to a reply
.TP
ARGSTRING
arguments to pass from
.B repl
to
.I CONVERTER
.TP
FILE
full path of message being replied to
.PD
.RE
The convert support is based on pseudoheaders of the form
.PP
.RS 5
.nf
.BI Nmh-mhbuild-file- TYPE : \0FILE
.BI Nmh-mhbuild-args- TYPE : \0ARGSTRING
.fi
.RE
.PP
in the draft.
For each such pseudoheader, mhbuild looks in the profile and
.I mhn.defaults
for the corresponding
.I TYPE
entry to find the converter that supports it:
.PP
.RS 5
.BI mhbuild-convert- TYPE : \0CONVERTER
.RE
.PP
It's a fatal error if no such entry is found for
.IR TYPE .
An empty entry, e.g.,
.PP
.RS 5
.B mhbuild-convert-text/html:
.RE
.PP
excludes parts of that
.I TYPE
from the draft.
.PP
The
.I mhn.defaults
file contains default
.B mhbuild-convert-text/html
and
.B mhbuild-convert-text/plain
entries.  Profile entries can be used to override corresponding
.I mhn.defaults
entries, as usual.  Text converters should limit text line lengths
to a maximum of 78 characters, and must limit them to a maximum of 998
characters, per RFC 5322 Sec.\& 2.1.1.
.PP
For each
.I TYPE
part in
.IR FILE ,
.B mhbuild
runs
.I CONVERTER ARGSTRING
on the content of the part.
Each part in
.I FILE
that has no corresponding TYPE entry in the profile or
.I mhn.defaults
is
excluded from the draft; the user can include them using mhbuild directives.
.PP
.B repl
inserts
.B Nmh-mhbuild-text/html:
and
.B Nmh-mhbuild-text/plain:
pseudoheaders in every draft.  The user can prevent insertion of
content parts of either of those types by putting corresponding empty
entries in their profile.
.PP
Only the highest precedence alternative with a supported
.I TYPE
of a multipart/alternative part is used.
.PP
mhn.defaults.sh selects the text/html-to-text/plain converter at
install time.  It includes
.B iconv
and
.BR par ,
or
.BR fmt ,
in the pipeline only if they are found.
.PP
Some content types require the addition of parameters to the
Content-Type header, such as \*(lqmethod=REPLY\*(rq
for text/calendar.
.B mhbuild
looks for a Content-Type header, followed
by a blank line, at the beginning of the converter output.
If one is found, it is used for the corresponding part in the reply draft.
.PP
The convert interface doesn't support different
.IR ARGSTRING s
or different converters for different parts of the same
.IR  TYPE .
That would require associating parts by part number with the
.IR ARGSTRING s
or converters.  Instead, that can be done (currently, without using
the convert support), with
.B mhbuild
directives as described below, e.g.,
.PP
.RS 5
#text/html; charset=utf-8 *8bit | mhstore -noverbose -part 42.7 -outfile - | w3m -dump -cols 64 -T text/html -O utf-8
.RE
.PP
The only way to mix convert pseudoheaders and
.B mhbuild
directives is to insert the directives before
.B mhbuild
is run, which is typically done by entering
.I mime
at the \*(lqWhat now?\*(rq prompt, or with an
.B \-editor mhbuild
switch.
.PP
These (optional) setup steps can make the convert support
easier to use:
.TP 5
1)
If the
.B par
program is installed on your system, it will be set by default
(in
.IR mhn.defaults )
to filter the converter output.  It helps to set the PARINIT
environment variable, as described in
.IR par (1).
.TP 5
2)
Add this line to your profile:
.IP "" 10
mhbuild-next: $EDITOR
.IP "" 5
assuming that your EDITOR environment variable is set; if not, replace
EDITOR with the name of your editor.  Without that profile entry, a
response of \*(lqe[dit]\*(rq at the What now? prompt will require
specification of your editor if an
.B \-editor mhbuild
switch is used.
.TP 5
3)
If using
.BR repl ,
source the Bourne-shell compatible functions in
.IR %docdir%/contrib/replaliases .
.br
That script also sets the PARINIT environment variable if it was not set.
.SS "Translating the Composition File"
.B mhbuild
is essentially a filter to aid in the composition of MIME messages.
.B mhbuild
will convert an
.B mhbuild
\*(lqcomposition file\*(rq into a valid MIME message.
An
.B mhbuild
\*(lqcomposition file\*(rq is just a file containing plain text that is
interspersed with various
.B mhbuild
directives.  When this file is processed by
.BR mhbuild ,
the various directives will be expanded to the appropriate content, and
will be encoded according to the MIME standards.
The resulting MIME message can then be sent by electronic mail.
.PP
The formal syntax for a
.B mhbuild
composition file is defined at the end of this document, but the ideas
behind this format are not complex.
Basically, the body contains one or more contents.  A content consists of
either a directive, indicated with a \*(lq#\*(rq as the first character
of a line; or, plaintext (one or more lines of text).
The continuation character, \*(lq\\\*(lq, may be used to enter a single
directive on more than one line, e.g.,
.PP
.RS 5
.nf
#image/png \\
    /home/foobar/junk/picture.png
.fi
.RE
.PP
There are five kinds of directives: \*(lqtype\*(rq directives, which
name the type and subtype of the content; \*(lqexternal-type\*(rq
directives, which also name the type and subtype of the content; the
\*(lqmessage\*(rq directive (#forw), which is used to forward one or
more messages; the \*(lqbegin\*(rq directive (#begin), which is
used to create a multipart content; and the \*(lqon/off/pop\*(rq
directives (#on, #off, #pop) which control whether any other
directives are honored at all.
.PP
The
.B \-directives
switch allows control over whether mhbuild will honor any of the
\*(lq#\*(rq-directives.
This can also be affected with the #on or #off directives, and #pop,
which restores the state of processing to that preceding the most recent
#on or #off.  (The #on, #off, and #pop directives are always honored,
of course.) This allows inclusion of plain text which looks like mhbuild
directives, without causing errors:
.PP
.RS 5
.nf
#off
#include <stdio.h>
puts("hello, world!");
#pop
.fi
.RE
.PP
Currently the stack depth for the #on/off/pop directives is 32.
.PP
The \*(lqtype\*(rq directive is used to directly specify the type and
subtype of a content.  You may only specify discrete types in this manner
(can't specify the types multipart or message with this directive).
You may optionally specify the name of a file containing the contents
in \*(lqnative\*(rq (decoded) format.
If this filename starts with the \*(lq|\*(rq character, then it represents
a command to execute whose output is captured accordingly.
For example,
.PP
.RS 5
.nf
#audio/basic |raw2audio -F < /usr/lib/sound/giggle.au
.fi
.RE
.PP
If a filename is not given,
.B mhbuild
will look for information in the
user's profile to determine how the different contents should be composed.
This is accomplished by consulting a composition string, and executing
it under
.BR /bin/sh ,
with the standard output set to the content.  If the
.B \-verbose
switch is given,
.B mhbuild
will echo any commands that are used to create contents in this way.
.PP
The composition string may contain the following escapes:
.PP
.RS 5
.PD 0
.IP %a
Insert parameters from directive
.IP %f
Insert filename containing content
.IP %F
%f, and stdout is not re-directed
.IP %s
Insert content subtype
.IP %%
Insert character %
.PD
.RE
.PP
First,
.B mhbuild
will look for an entry of the form:
.PP
.RS 5
.BI mhbuild-compose- type / subtype
.RE
.PP
to determine the command to use to compose the content.
If this isn't found,
.B mhbuild
will look for an entry of the form:
.PP
.RS 5
.BI mhbuild-compose- type
.RE
.PP
to determine the composition command.  If this isn't found,
.B mhbuild
will complain.
.PP
An example entry might be:
.PP
.RS 5
mhbuild-compose-audio/basic: record | raw2audio -F
.RE
.PP
Because commands like these will vary, depending on the display
environment used for login, composition strings for different
contents should probably be put in the file specified by the
MHBUILD environment variable, instead of directly in your
user profile.
.PP
The \*(lqexternal-type\*(rq directives are used to provide a MIME
reference to a content, rather than enclosing the contents itself
(for instance, by specifying an ftp site).
Hence, instead of providing a filename as with the type directives,
external-parameters are supplied.
These look like regular parameters, so they must be separated accordingly.
For example,
.PP
.RS 5
.nf
#@application/octet-stream; \\
    type=tar; \\
    conversions=compress \\
    [this is the nmh distribution] \\
    {attachment; filename="nmh.tar.gz"} \\
    name="nmh.tar.gz"; \\
    directory="/pub/nmh"; \\
    site="ftp.math.gatech.edu"; \\
    access-type=anon-ftp; \\
    mode="image"
.fi
.RE
.PP
You must give a description string to separate the content parameters
from the external-parameters (although this string may be empty).
This description string is specified by enclosing it within
\*(lq[]\*(rq.
A disposition string, to appear in a
\*(lqContent-Disposition\*(rq header, may appear in the optional
\*(lq{}\*(rq.
.PP
These parameters are of the form:
.PP
.RS 5
.nf
.ta \w'access-type=  'u
access-type=	usually \*(lqanon-ftp\*(rq, \*(lqmail-server\*(rq, or \*(lqurl\*(rq
name=	filename
permission=	read-only or read-write
site=	hostname
directory=	directoryname (optional)
mode=	usually \*(lqascii\*(rq or \*(lqimage\*(rq (optional)
size=	number of octets
server=	mailbox
subject=	subject to send
body=	command to send for retrieval
url=	URL of content
.fi
.RE
.PP
A minimum \*(lqexternal\-type\*(rq directive for the
.B url
.I access-type
would be as follows:
.PP
.RS 3
.nf
#@application/octet-stream [] access-type=url; \\
  url="http://download.savannah.gnu.org/releases/nmh/nmh-1.5.tar.gz"
.fi
.RE
.PP
Any long URLs will be wrapped according to RFC 2231 rules.
.PP
The \*(lqmessage\*(rq directive (#forw) is used to specify a message or
group of messages to include.  You may optionally specify the name of
the folder and which messages are to be forwarded.
If a folder is not given, it defaults to the current folder.
Similarly, if a message is not given, it defaults to the current message.
Hence, the message directive is similar to the
.B forw
command, except that the former uses the MIME rules for encapsulation
rather than those specified in RFC 934.
For example,
.PP
.RS 5
.nf
#forw +inbox 42 43 99
.fi
.RE
.PP
If you include a single message, it will be included directly as a content
of type \*(lqmessage/rfc822\*(rq.
If you include more than one message, then
.B mhbuild
will add a content of type \*(lqmultipart/digest\*(rq
and include each message as a subpart of this content.
.PP
If you are using this directive to include more than one message, you
may use the
.B \-rfc934mode
switch.  This switch will indicate that
.B mhbuild
should attempt to utilize the MIME encapsulation rules in such a way
that the \*(lqmultipart/digest\*(rq that is created is (mostly) compatible
with the encapsulation specified in RFC 934.
If given, then RFC 934 compliant user-agents should be able to burst the
message on reception\0--\0providing that the messages being encapsulated
do not contain encapsulated messages themselves.
The drawback of this approach is that the encapsulations are generated by
placing an extra newline at the end of the body of each message.
.PP
The \*(lqbegin\*(rq directive is used to create a multipart content.
When using the \*(lqbegin\*(rq directive, you must specify at least one
content between the begin and end pairs.
.PP
.RS 5
.nf
#begin
This will be a multipart with only one part.
#end
.fi
.RE
.PP
If you use multiple directives in a composition draft,
.B mhbuild
will automatically encapsulate them inside a multipart content.
Therefore the \*(lqbegin\*(rq directive is only necessary if you wish to
use nested multiparts, or create a multipart message containing only one part.
.PP
For all of these directives, the user may include a brief description
of the content between the \*(lq[\*(rq character and the \*(lq]\*(rq
character.
This description will be copied into the \*(lqContent-Description\*(rq header
when the directive is processed.
.PP
.RS 5
.nf
#forw [important mail from Bob] +bob 1 2 3 4 5
.fi
.RE
.PP
Similarly, a disposition string may optionally be provided between
\*(lq{\*(rq and \*(lq}\*(rq characters; it will be copied into the
\*(lqContent-Disposition\*(rq header when the directive is processed.
If a disposition string is provided that does not contain a filename
parameter, and a filename is provided in the directive, it will be
added to the \*(lqContent-Disposition\*(rq header.
For example, the following directive:
.PP
.RS 5
.nf
#text/plain; charset=iso-8859-1 <>{attachment} /tmp/summary.txt
.fi
.RE
.PP
creates these message part headers:
.PP
.RS 5
.nf
Content-Type: text/plain; charset="iso-8859-1"
Content-Disposition: attachment; filename="summary.txt"
.fi
.RE
.PP
By default,
.B mhbuild
will generate a unique \*(lqContent-ID:\*(rq for each directive,
corresponding to each message part; however, the user may override
this by defining the ID using the \*(lq<\*(rq and \*(lq>\*(rq
characters.  The
.B \-nocontentid
switch suppresses creation of all \*(lqContent-ID:\*(rq headers,
even in the top level of the message.
.PP
Normally
.B mhbuild
will choose an appropriate Content\-Transfer\-Encoding based on the content
and the MIME Content\-Type.
However, you can override that in an
.B mhbuild
directive by specifying \*(lq*\*(rq and the encoding.
Acceptable encoding values are \*(lq8bit\*(rq, \*(lqqp\*(rq
(for quoted-printable), and \*(lqb64\*(rq (for base64 encoding).
It should be noted that undesired results may occur if 8bit or quoted-printable
is selected for binary content, due to the translation between Unix line endings
and the line endings use by the mail transport system.
.PP
In addition to the various directives, plaintext can be present.
Plaintext is gathered, until a directive is found or the draft is
exhausted, and this is made to form a text content.
If the plaintext must contain a \*(lq#\*(rq at the beginning of a line,
simply double it,
e.g.,
.PP
.RS 5
##when sent, this line will start with only one #
.RE
.PP
If you want to end the plaintext prior to a directive, e.g., to have two
plaintext contents adjacent, simply insert a line containing a single
\*(lq#\*(rq character, e.g.,
.PP
.RS 5
.nf
this is the first content
#
and this is the second
.fi
.RE
.PP
Finally, if the plaintext starts with a line of the form:
.PP
.RS 5
Content-Description: text
.RE
.PP
then this will be used to describe the plaintext content.
You MUST follow this line with a blank line before starting your text.
.PP
By default, plaintext is captured as a text/plain content.
You can override this by starting the plaintext with \*(lq#<\*(rq followed
by a content-type specification.  For example, e.g.,
.PP
.RS 5
.nf
#<text/enriched
this content will be tagged as text/enriched
#
and this content will be tagged as text/plain
#
#<application/x-patch [this is a patch]
and this content will be tagged as application/x-patch
.fi
.RE
.PP
Note that if you use the \*(lq#<\*(rq plaintext-form, then the
content-description must be on the same line which identifies the content
type of the plaintext.
.PP
When composing a text content, you may indicate the relevant character
set by adding the \*(lqcharset\*(rq parameter to the directive.
.PP
.RS 5
#<text/plain; charset=iso-8859-5
.RE

NEWS
# NEWS - Release notes for nmh 1.7
#

Welcome to nmh, the new version of the classic MH mail handling system.
It's been over three years since the last release of nmh, and there have
been a number of significant changes since the last release.  Long-time
MH and nmh uses should read careful the NOTEABLE CHANGES section, as there
are some significant changes to nmh behavior.  Otherwise, please see the
README and INSTALL files for help on getting started with nmh.

For news of future releases, subscribe to the low-volume
    https://lists.nongnu.org/mailman/listinfo/nmh-announce

---------------
NOTABLE CHANGES
---------------

The largest notable changes in the 1.7 release are:

- Complete unification of network security support.  All network protocols
  (currently, POP and SMTP) have been refactored to use a common set of
  security routines.  This means all protocols support all SASL mechanisms
  (via the Cyrus-SASL library) and TLS.  TLS support has been strengthened
  to perform certificate name validation and to require TLS 1.1 as a
  minimum protocol.  Also, all protocols can make use of the OAuth2/XOAUTH
  SASL mechanism, which is supported by Gmail.
- send(1) now supports adding switches to post(8) based on the address or
  domain of the email address in the From: header; this more easily allows
  users to support multiple identities.
- A generic facility for passing arguments to filter programs in repl(1)
  by use of the -convertargs switch.
- Native support for the manipulation of iCalendar requests; see mhical(1)
  for more details.

------------
NEW FEATURES
------------

The following are new features for the 1.7 release of nmh:

- When building from source, configure will derive ${prefix} from an existing
  nmh installation if it finds one in your $PATH.
- Added welcome message when nmh detects that its version changed.
- The default locations for configuration files and support binaries
  have been changed.  Configuration files now install into ${sysconfdir}/nmh,
  and support binaries are placed in ${libexecdir}/nmh.  If you are upgrading
  an existing installation you should look for old configuration files in
  ${sysconfdir} and merge any local customizations into the new files in
  ${sysconfdir}/nmh, then remove the old files.  ${libdir} will also contain
  obsolete support programs that should be removed.
- All TLS connections now perform certificate validation (including hostname
  matching) by default; can be disabled on a per-application basis.
- post now defaults to port 587 on 'smtp' message submission.
- A value of 0 for the width switch of scan(1), inc(1), ap(1), dp(1),
  fmttest(1), and mhl(1) now means as many characters as the format
  engine can produce [Bug #15274].  That amount is limited by internal
  buffers.
- If a component has trailing whitespace, e.g., body:component="> ",
  mhl now trims that whitespace off when filtering blank text lines.
- An "rtrim" flag has been added to mhl to remove any trailing
  whitespace from filtered text lines.  A corresponding "nortrim" flag
  has also been added.
- Added getmymbox and getmyaddr mh-format(5) function escapes.
- New -[no]changecur, -fixtype, -decodetypes, and -[no]crlflinebreaks switches
  have been added to mhfixmsg(1).
- mhfixmsg now removes an extraneous trailing semicolon from header
  parameter lists.
- Added -convertargs switch to repl(1), to pass arguments to programs
  specified in the user's profile or mhn.defaults to convert message
  content.
- Added mhical(1), to display, reply to, and cancel iCalendar (RFC 5545)
  event requests.
- Added multiply mh-format(5) function.
- "mhparam bindir" prints the path to the directory containing the public
  executables (${bindir}).
- New "-prefer" switch for mhshow (and mhlist and mhshow), to allow specifying
  the preferred content types to show, if present in a multipart alternative.
- mh-format now has %(kilo) and %(kibi) functions, to allow printing
  numbers with SI or IEC quantities, e.g. "10K", "2.3Mi".
- Support for the -sendmail flag to send/post to change the sendmail
  binary when using the sendmail/pipe MTS.
- Added support to send(1) to specify switches to post(1) based on address or
  domain name in From: header line in message draft.
- post(8) -snoop now attempts to decode base64-encoded SMTP traffic.
- folder(1) -nocreate now prints a warning message for a non-existent folder.
- mhfixmsg(1) now allows -decodetext binary, though 8bit is still the default.
- inc(1) and msgchk(1) now support TLS encryption natively.
- All network protocols support the XOAUTH authentication mechanism.
- Support for SMTPUTF8 (RFC 6531) has been added.  mhshow(1) already supported
  RFC 6532, assuming all 8-bit message header field bodies are UTF-8 and use
  of a UTF-8 locale.
- mhfixmsg now replaces RFC 2047 encoding with RFC 2231 encoding of name and
  filename parameters in Content-Type and Content-Disposition headers,
  respectively.
- If a message body contains 8-bit bytes, post(8) uses SMTP 8BITMIME if the
  server supports it.  If not, post fails with a message to the user to
  encode the message for 7-bit transport.
- Fewer lseek(2)s will be used when reading headers in the common case.
- ./configure's --enable-debug has been removed;  it did nothing.
- configure now defaults to enabling each of TLS and Cyrus SASL if the
  necessary headers and libraries are found.
- Moved build_nmh to top-level directory.
- Better error reporting for connections to network services.

---------
BUG FIXES
---------

- The format scanner no longer subtracts 1 from the width.  This has the
  effect of no longer counting the trailing newline in the output of
  scan(1), inc(1), and the other programs that rely on the format scanner.
- The first character of some very short (less than 4 characters) message
  bodies is no longer dropped.
- Single-character headers can be reliably formatted, etc., instead of
  apparently being missing.

uip/mhbuild.c
/* mhbuild.c -- expand/translate MIME composition files
 *
 * This code is Copyright (c) 2002, by the authors of nmh.  See the
 * COPYRIGHT file in the root directory of the nmh distribution for
 * complete copyright information.
 */

#include "h/mh.h"
#include "sbr/getarguments.h"
#include "sbr/smatch.h"
#include "sbr/m_backup.h"
#include "sbr/context_find.h"
#include "sbr/readconfig.h"
#include "sbr/ambigsw.h"
#include "sbr/path.h"
#include "sbr/print_version.h"
#include "sbr/print_help.h"
#include "sbr/error.h"
#include <fcntl.h>
#include "h/mts.h"
#include "h/tws.h"
#include "h/mime.h"
#include "h/mhparse.h"
#include "h/done.h"
#include "h/utils.h"
#include "sbr/m_maildir.h"
#include "sbr/m_mktemp.h"
#include "mhfree.h"
#include "mhoutsbr.h"

#define MHBUILD_SWITCHES \
    X("auto", 0, AUTOSW) \
    X("noauto", 0, NAUTOSW) \
    X("check", -5, CHECKSW) \
    X("nocheck", -7, NCHECKSW) \
    X("directives", 0, DIRECTIVES) \
    X("nodirectives", 0, NDIRECTIVES) \
    X("headers", 0, HEADSW) \
    X("noheaders", 0, NHEADSW) \
    X("list", 0, LISTSW) \
    X("nolist", 0, NLISTSW) \
    X("realsize", 0, SIZESW) \
    X("norealsize", 0, NSIZESW) \
    X("rfc934mode", 0, RFC934SW) \
    X("norfc934mode", 0, NRFC934SW) \
    X("verbose", 0, VERBSW) \
    X("noverbose", 0, NVERBSW) \
    X("disposition", 0, DISPOSW) \
    X("nodisposition", 0, NDISPOSW) \
    X("contentid", 0, CONTENTIDSW) \
    X("nocontentid", 0, NCONTENTIDSW) \
    X("headerencoding encoding-algorithm", 0, HEADERENCSW) \
    X("autoheaderencoding", 0, AUTOHEADERENCSW) \
    X("maxunencoded", 0, MAXUNENCSW) \
    X("version", 0, VERSIONSW) \
    X("help", 0, HELPSW) \
    X("debug", -5, DEBUGSW) \
    X("dist", 0, DISTSW) \

#define X(sw, minchars, id) id,
DEFINE_SWITCH_ENUM(MHBUILD);
#undef X

#define X(sw, minchars, id) { sw, minchars, id },
DEFINE_SWITCH_ARRAY(MHBUILD, switches);
#undef X

/* utf-8 is for Email Address Internationalization, using SMTPUTF8. */
#define MIMEENCODING_SWITCHES \
    X("base64", 0, BASE64SW) \
    X("quoted-printable", 0, QUOTEDPRINTSW) \
    X("utf-8", 0, UTF8SW) \

#define X(sw, minchars, id) id,
DEFINE_SWITCH_ENUM(MIMEENCODING);
#undef X

#define X(sw, minchars, id) { sw, minchars, id },
DEFINE_SWITCH_ARRAY(MIMEENCODING, encodingswitches);
#undef X

int debugsw = 0;

bool listsw;
bool rfc934sw;
bool contentidsw = true;

/*
 * Temporary files
 */
static char infile[BUFSIZ];
static char outfile[BUFSIZ];


int
main (int argc, char **argv)
{
    bool sizesw = true;
    bool headsw = true;
    bool directives = true;
    bool autobuild = false;
    bool dist = false;
    bool verbosw = false;
    bool dispo = false;
    size_t maxunencoded = MAXTEXTPERLN;
    char *cp, buf[BUFSIZ];
    char buffer[BUFSIZ], *compfile = NULL;
    char **argp, **arguments;
    CT ct, cts[2];
    FILE *fp = NULL;
    FILE *fp_out = NULL;
    int header_encoding = CE_UNKNOWN;
    size_t n;

    if (nmh_init(argv[0], true, false)) { return 1; }

    arguments = getarguments (invo_name, argc, argv, 1);
    argp = arguments;

    while ((cp = *argp++)) {
	if (cp[0] == '-' && cp[1] == '\0') {
	    if (compfile)
		die("cannot specify both standard input and a file");
            compfile = cp;
	    listsw = false;	/* turn off -list if using standard in/out */
	    verbosw = false;	/* turn off -verbose listings */
	    break;
	}
	if (*cp == '-') {
	    switch (smatch (++cp, switches)) {
	    case AMBIGSW: 
		ambigsw (cp, switches);
		done (1);
	    case UNKWNSW: 
		die("-%s unknown", cp);

	    case HELPSW: 
		snprintf (buf, sizeof(buf), "%s [switches] file", invo_name);
		print_help (buf, switches, 1);
		done (0);
	    case VERSIONSW:
		print_version(invo_name);
		done (0);

	    case AUTOSW:
		/* -auto implies -nodirectives */
		autobuild = true;
		directives = false;
		continue;
	    case NAUTOSW:
		/*
		 * We're turning directives back on since this is likely here
		 * to override a profile entry
		 */
		autobuild = false;
		directives = true;
		continue;

	    case CHECKSW:
	    case NCHECKSW:
		/* Currently a NOP */
		continue;

	    case HEADSW:
		headsw = true;
		continue;
	    case NHEADSW:
		headsw = false;
		continue;

	    case DIRECTIVES:
		directives = true;
		continue;
	    case NDIRECTIVES:
		directives = false;
		continue;

	    case LISTSW:
		listsw = true;
		continue;
	    case NLISTSW:
		listsw = false;
		continue;

	    case RFC934SW:
		rfc934sw = true;
		continue;
	    case NRFC934SW:
		rfc934sw = false;
		continue;

	    case SIZESW:
		sizesw = true;
		continue;
	    case NSIZESW:
		sizesw = false;
		continue;

	    case CONTENTIDSW:
		contentidsw = true;
		continue;
	    case NCONTENTIDSW:
		contentidsw = false;
		continue;

	    case HEADERENCSW: {
		int encoding;

		if (!(cp = *argp++) || *cp == '-')
		    die("missing argument to %s", argp[-2]);
		switch (encoding = smatch (cp, encodingswitches)) {
		case AMBIGSW:
		    ambigsw (cp, encodingswitches);
		    done (1);
		case UNKWNSW:
		    die("%s unknown encoding algorithm", cp);
		case BASE64SW:
		    header_encoding = CE_BASE64;
		    break;
		case QUOTEDPRINTSW:
		    header_encoding = CE_QUOTED;
		    break;
		case UTF8SW:
		    header_encoding = CE_8BIT;
		    break;
		default:
		    die("Internal error: algorithm %s", cp);
		}
		continue;
	    }

	    case AUTOHEADERENCSW:
		header_encoding = CE_UNKNOWN;
		continue;

	    case MAXUNENCSW:
		if (!(cp = *argp++) || *cp == '-')
		    die("missing argument to %s", argp[-2]);
		if ((maxunencoded = atoi(cp)) < 1)
		    die("Invalid argument for %s: %s", argp[-2], cp);
		if (maxunencoded > 998)
		    die("limit of -maxunencoded is 998");
		continue;

	    case VERBSW: 
		verbosw = true;
		continue;
	    case NVERBSW: 
		verbosw = false;
		continue;
	    case DISPOSW:
		dispo = true;
		continue;
	    case NDISPOSW:
		dispo = false;
		continue;
	    case DEBUGSW:
		debugsw = 1;
		continue;
	    case DISTSW:
		dist = true;
		continue;
	    }
	}
	if (compfile)
	    die("only one composition file allowed");
        compfile = cp;
    }

    /*
     * Check if we've specified an additional profile
     */
    if ((cp = getenv ("MHBUILD"))) {
	if ((fp = fopen (cp, "r"))) {
	    readconfig(NULL, fp, cp, 0);
	    fclose (fp);
	} else {
	    admonish ("", "unable to read $MHBUILD profile (%s)", cp);
	}
    }

    /*
     * Read the standard profile setup
     */
    if ((fp = fopen (cp = etcpath ("mhn.defaults"), "r"))) {
	readconfig(NULL, fp, cp, 0);
	fclose (fp);
    }

    if (!context_find ("path"))
	free (path ("./", TFOLDER));

    /* Check if we have a file to process */
    if (!compfile)
	die("need to specify a %s composition file", invo_name);

    /*
     * Process the composition file from standard input.
     */
    if (compfile[0] == '-' && compfile[1] == '\0') {
	if ((cp = m_mktemp2(NULL, invo_name, NULL, &fp)) == NULL) {
	    die("unable to create temporary file in %s",
		  get_temp_dir());
	}
	strncpy (infile, cp, sizeof(infile));

	/* copy standard input to temporary file */
	while ((n = fread(buffer, 1, sizeof(buffer), stdin)) > 0) {
	    if (fwrite(buffer, 1, n, fp) != n) {
		die("error copying to temporary file");
	    }
	}
	fclose (fp);

	/* build the content structures for MIME message */
	ct = build_mime (infile, autobuild, dist, directives, header_encoding,
			 maxunencoded, verbosw);

	/*
	 * If ct == NULL, that means that -auto was set and a MIME version
	 * header was already seen.  Just use the input file as the output
	 */

	if (!ct) {
	    if (! (fp = fopen(infile, "r"))) {
		die("Unable to open %s for reading", infile);
	    }
	    while ((n = fread(buffer, 1, sizeof(buffer), fp)) > 0) {
		if (fwrite(buffer, 1, n, stdout) != n) {
		    die("error copying %s to stdout", infile);
		}
	    }
	} else {
	    /* output the message */
	    output_message_fp (ct, stdout, NULL);
	    free_content (ct);
	}

	done (0);
    }

    /*
     * Process the composition file from a file.
     */

    /* build the content structures for MIME message */
    ct = build_mime (compfile, autobuild, dist, directives, header_encoding,
		     maxunencoded, verbosw);

    /*
     * If ct == NULL, that means -auto was set and we found a MIME version
     * header.  Simply exit and do nothing.
     */

    if (! ct)
	done(0);

    cts[0] = ct;
    cts[1] = NULL;

    /* output MIME message to this temporary file */
    if ((cp = m_mktemp2(compfile, invo_name, NULL, &fp_out)) == NULL) {
	die("unable to create temporary file");
    }
    strncpy(outfile, cp, sizeof(outfile));

    /* output the message */
    output_message_fp (ct, fp_out, outfile);
    fclose(fp_out);

    /*
     * List the message info
     */
    if (listsw)
	list_all_messages (cts, headsw, sizesw, verbosw, debugsw, dispo);

    /* Rename composition draft */
    snprintf (buffer, sizeof(buffer), "%s.orig", m_backup (compfile));
    if (rename (compfile, buffer) == NOTOK) {
	adios (compfile, "unable to rename comp draft %s to", buffer);
    }

    /* Rename output file to take its place */
    if (rename (outfile, compfile) == NOTOK) {
	advise (outfile, "unable to rename output %s to", compfile);
	rename (buffer, compfile);
	done (1);
    }
    /* Remove from atexit(3) list of files to unlink. */
    if (!(m_unlink(outfile) == -1 && errno == ENOENT)) {
        adios(outfile, "file exists after rename:");
    }

    free_content (ct);
    done (0);
    return 1;
}
